File: content/tracks/dsa-python/concepts/17-tries/tries_test.py
import sys
from stub import Trie, search_suggestions

passed = 0
failed = 0


def test(name: str, condition: bool, detail: str = "") -> None:
    global passed, failed
    if condition:
        print(f"  PASS: {name}")
        passed += 1
    else:
        msg = f" — {detail}" if detail else ""
        print(f"  FAIL: {name}{msg}")
        failed += 1


# Trie
t = Trie()
t.insert("apple")
test("trie: search exact word", t.search("apple") is True)
test("trie: prefix not a word", t.search("app") is False)
test("trie: starts_with valid prefix", t.starts_with("app") is True)
test("trie: starts_with full word", t.starts_with("apple") is True)
test("trie: starts_with missing prefix", t.starts_with("xyz") is False)
t.insert("app")
test("trie: search after inserting prefix", t.search("app") is True)
test("trie: apple still found", t.search("apple") is True)
test("trie: shorter prefix 'ap' not a word", t.search("ap") is False)

t2 = Trie()
t2.insert("")
test("trie: empty string insert+search", t2.search("") is True)
test("trie: empty prefix always starts_with", t2.starts_with("") is True)

t3 = Trie()
for w in ["bat", "ball", "band", "can", "cat"]:
    t3.insert(w)
test("trie: multi-word search bat", t3.search("bat") is True)
test("trie: multi-word prefix ba", t3.starts_with("ba") is True)
test("trie: multi-word search ba not word", t3.search("ba") is False)
test("trie: starts_with cap false", t3.starts_with("cap") is False)

# search_suggestions
r = search_suggestions(["mobile","mouse","moneypot","monitor","mousepad"], "mouse")
test("suggestions: prefix m", r[0] == ["mobile","moneypot","monitor"], f"got {r[0]}")
test("suggestions: prefix mo", r[1] == ["mobile","moneypot","monitor"], f"got {r[1]}")
test("suggestions: prefix mou", r[2] == ["mouse","mousepad"], f"got {r[2]}")
test("suggestions: prefix mous", r[3] == ["mouse","mousepad"], f"got {r[3]}")
test("suggestions: prefix mouse", r[4] == ["mouse","mousepad"], f"got {r[4]}")

r2 = search_suggestions(["havana"], "havana")
test("suggestions: single product", all(row == ["havana"] for row in r2), f"got {r2}")

r3 = search_suggestions(["bags","baggage","banner","box","cloths"], "bags")
test("suggestions: bags prefix b", r3[0] == ["baggage","bags","banner"], f"got {r3[0]}")
test("suggestions: bags prefix ba", r3[1] == ["baggage","bags","banner"], f"got {r3[1]}")
test("suggestions: bags prefix bag", r3[2] == ["baggage","bags"], f"got {r3[2]}")
test("suggestions: bags prefix bags", r3[3] == ["bags"], f"got {r3[3]}")

print(f"\n{passed} passed, {failed} failed")
if failed > 0:
    sys.exit(1)

File: content/tracks/dsa-typescript/concepts/17-tries/tries_test.ts
import { Trie, searchSuggestions } from "./stub";

let passed = 0;
let failed = 0;

function test(name: string, fn: () => void): void {
  try {
    fn();
    console.log(`  PASS: ${name}`);
    passed++;
  } catch (e: unknown) {
    const msg = e instanceof Error ? e.message : String(e);
    console.log(`  FAIL: ${name} — ${msg}`);
    failed++;
  }
}

function assertEqual<T>(actual: T, expected: T): void {
  const a = JSON.stringify(actual);
  const b = JSON.stringify(expected);
  if (a !== b) throw new Error(`expected ${b}, got ${a}`);
}

// --- Trie ---

test("Trie: search inserted word", () => {
  const t = new Trie();
  t.insert("apple");
  assertEqual(t.search("apple"), true);
});

test("Trie: search returns false for prefix only", () => {
  const t = new Trie();
  t.insert("apple");
  assertEqual(t.search("app"), false);
});

test("Trie: search returns false for non-inserted word", () => {
  const t = new Trie();
  t.insert("apple");
  assertEqual(t.search("banana"), false);
});

test("Trie: search empty string not inserted", () => {
  const t = new Trie();
  assertEqual(t.search(""), false);
});

test("Trie: search empty string after inserting it", () => {
  const t = new Trie();
  t.insert("");
  assertEqual(t.search(""), true);
});

test("Trie: multiple words, each found independently", () => {
  const t = new Trie();
  ["cat", "car", "card", "care"].forEach((w) => t.insert(w));
  assertEqual(t.search("car"), true);
  assertEqual(t.search("card"), true);
  assertEqual(t.search("cat"), true);
});

test("Trie: shorter word is not found if only longer was inserted", () => {
  const t = new Trie();
  t.insert("card");
  assertEqual(t.search("car"), false);
});

test("Trie: startsWith returns true for valid prefix", () => {
  const t = new Trie();
  t.insert("apple");
  assertEqual(t.startsWith("app"), true);
});

test("Trie: startsWith returns true for full word", () => {
  const t = new Trie();
  t.insert("apple");
  assertEqual(t.startsWith("apple"), true);
});

test("Trie: startsWith returns false for absent prefix", () => {
  const t = new Trie();
  t.insert("apple");
  assertEqual(t.startsWith("ban"), false);
});

test("Trie: startsWith empty string is always true after any insert", () => {
  const t = new Trie();
  t.insert("apple");
  assertEqual(t.startsWith(""), true);
});

test("Trie: same word inserted twice, search still true", () => {
  const t = new Trie();
  t.insert("hello");
  t.insert("hello");
  assertEqual(t.search("hello"), true);
});

// --- searchSuggestions ---

test("searchSuggestions: standard case", () => {
  const products = ["mobile","mouse","moneypot","monitor","mousepad"];
  const result = searchSuggestions(products, "mouse");
  assertEqual(result, [
    ["mobile","moneypot","monitor"],
    ["mobile","moneypot","monitor"],
    ["mouse","mousepad"],
    ["mouse","mousepad"],
    ["mouse","mousepad"],
  ]);
});

test("searchSuggestions: no matches after first char", () => {
  const result = searchSuggestions(["apple","app"], "xyz");
  assertEqual(result, [[], [], []]);
});

test("searchSuggestions: exactly 3 suggestions at one level", () => {
  const products = ["a","ab","abc","abcd"];
  const result = searchSuggestions(products, "abc");
  // prefix "a" → up to 3: ["a","ab","abc"]
  assertEqual(result[0], ["a","ab","abc"]);
  // prefix "ab" → up to 3: ["ab","abc","abcd"]
  assertEqual(result[1], ["ab","abc","abcd"]);
  // prefix "abc" → up to 3: ["abc","abcd"]
  assertEqual(result[2], ["abc","abcd"]);
});

test("searchSuggestions: single product", () => {
  const result = searchSuggestions(["hello"], "hel");
  assertEqual(result, [["hello"], ["hello"], ["hello"]]);
});

test("searchSuggestions: empty product list", () => {
  const result = searchSuggestions([], "abc");
  assertEqual(result, [[], [], []]);
});

console.log(`\n${passed} passed, ${failed} failed`);
if (failed > 0) process.exit(1);
